Template Pipeline: content projection

dylhunn opened the pull request:
Add support for content projection to the template pipeline, including `projection`, `projectionDef`, `ngContentSelectors`, and `ngProjectAs`. See individual commits for details.

dylhunn committed:
refactor(compiler): Add an explicit `controlFlow` property to template ops

The new built-in control flow design includes calls to the `template` instruction with fewer arguments. This was previously handled implicitly, but it's more extensible to add an explicit flag to the template op to handle this case.

dylhunn committed:
refactor(compiler): Support content projection in template pipeline

Supporting content projection requires us to emit three new kinds of output:

1. An `ngContentSelectors` field on the component metadata, which points to an array in the constant pool with all of the `select` attributes from `<ng-content>` elements.
2. One `projectionDef` instruction at the beginning of each root view template function for a component. That `projectionDef` points to a constant pool expression, which contains *parsed* selectors for all `<ng-content>` elements in the root's entire view tree.
3. A `projection` instruction for each `<ng-content>` slot in the view tree. These each get a data slot, a monotonically increasing "content slot", and a pointer to the tag's attributes in the component const array.

We support the first two features entirely within a new compilation phase.

The third feature, collection of processed attributes, is a bit trickier. We now treat `<ng-content>` tags as element-like ops, and use the normal attribute ingestion pipeline to process any attributes, and assign the appropriate `ConstIndex`.

**Note**: We also split up a number of the tests into two expectations files, one for the view functions, and one for other const listerals from the constant pool. This is because `TemplateDefinitionBuilder` emits the literals in a quirky order (mixed in with the view functions) due to how it lazily generates view functions. Our eager ordering is totally different, but by splitting the expectations, we can still share the same tests with `TemplateDefinitionBuilder`.

dylhunn committed:
refactor(compiler): Split up the const collection phase

The const collection phase is getting too large. We simplify it by breaking out i18n const collection into a separate phase.

dylhunn committed:
refactor(compiler): Support `ngProjectAs` attributes

Content project allows the content to specify its own selector for matching against content projection slots, using the `ngProjectAs` special attribute. We can now treat this attribue specially, and generate the appropriate flag in the consts array, followed by the parsed CSS selector.

dylhunn committed:
refactor(compiler): Enable some additional passing tests for template pipeline

A couple more tests pertaining to animations became passing at some point recently, as well as a few form other assorted areas.

alxhub commented on packages/compiler/src/template/pipeline/ir/src/ops/create.ts:
```diff
@@ -170,19 +171,25 @@ export interface TemplateOp extends ElementOpBase {
+  /**
+   * Whether or not this template was automatically created for built-in control flow.
+   */
+  controlFlow: boolean;
```
Rather than have a flag with a very specific purpose, could we not make `tag` nullable instead?

That way, we encode the actual kind of thing we're creating, a "template without a tag", rather than something specific to control flow.

dylhunn replied:
I made an effort here, but it has rather messy impacts on all the rest of the element/container code, since it's generally assumed that `tag` is present. Added a TODO to consider a deeper refactor, in which we separate control flow into a different op type.

alxhub commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -180,6 +180,23 @@ function ingestTemplate(unit: ViewCompilationUnit, tmpl: t.Template): void {
+  for (const attr of content.attributes) {
+    // Attributes named 'select' are specifically excluded, because they control which content
+    // matches as a property of the `projection`, and are not a plain attribute.
+    if (attr.name.toLowerCase() !== 'select') {
```
Rather than encode this behavior during ingest, we could add a phase which drops these bindings later. Or is there a specific reason why this doesn't work?

dylhunn replied:
Done.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/generate_projection_def.ts:
```diff
@@ -0,0 +1,53 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {parseSelectorToR3Selector} from '../../../../core';
```
This import looks wrong - we should import it from the actual file it's defined in.

dylhunn replied:
This actually is exported from a file called "core.ts"!

alxhub commented on packages/compiler/src/template/pipeline/src/phases/i18n_const_collection.ts:
```diff
@@ -0,0 +1,35 @@
+ */
+export function phaseI18nConstCollection(job: ComponentCompilationJob): void {
+  // Serialize the extracted messages into the const array.
+  const messageConstIndices: {[id: ir.XrefId]: ir.ConstIndex} = {};
```
Prefer `Map` to object maps.

dylhunn replied:
This code was just moved out of `const_collection`, but I have added a TODO to fix it in a future i18n commit.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/const_collection.ts:
```diff
@@ -105,6 +106,18 @@ class ElementAttributes {
+      if (value === null || !(value instanceof o.LiteralExpr)) {
+        throw Error('ngProjectAs must have a string literal value');
+      }
+      this.projectAs = value.value as string;
```
We should check `typeof value.value` instead of casting here.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/const_collection.ts:
```diff
@@ -105,6 +106,18 @@ class ElementAttributes {
     }
     this.known.add(name);
+    if (name === 'ngProjectAs') {
+      // TODO: Perhaps we should consider ingesting raw attribute strings instead of expressions, in
```
It's tempting, but actually this is the right behavior. Otherwise, we'd have to special case _all_ binding logic to handle raw strings as well as expressions.

alxhub reviewed (approved):
Nice work! A few minor points and design questions, but overall this looks great!

dylhunn committed:
refactor(compiler): Clean up a few messy content projection details

Pull out exclusion of `select` attributes on content nodes into a phase, and clean up a couple other details of content projection.

dylhunn commented:
caretaker: windows is flaky; no impact on g3
